lib/cogctl/actions/groups/add.ex

```diff
@@ -1,61 +1,41 @@
 defmodule Cogctl.Actions.Groups.Add do
   use Cogctl.Action, "groups add"
+
   alias Cogctl.Actions.Groups
+  alias CogApi.Resources.User
+  alias CogApi.HTTP.Client
 
   def option_spec do
     [{:group, :undefined, :undefined, {:string, :undefined}, 'Group name (required)'},
-     {:user_to_add, :undefined, 'user', {:string, :undefined}, 'Username of user to add'},
-     {:group_to_add, :undefined, 'group', {:string, :undefined}, 'Name of group to add'}]
-  end
-
-  def run(options, _args, _config, endpoint) do
-    group = :proplists.get_value(:group, options)
-    user_to_add = :proplists.get_value(:user_to_add, options)
-    group_to_add = :proplists.get_value(:group_to_add, options)
-
-    with_authentication(endpoint,
-                        &do_add(&1, group, user_to_add, group_to_add))
-  end
-
-  defp do_add(_endpoint, :undefined, _user_to_add, _group_to_add) do
-    display_arguments_error
+     {:email, :undefined, 'email', {:string, :undefined}, 'User email address (required)'}]
   end
 
-  defp do_add(_endpoint, _group_name, :undefined, :undefined) do
-    display_arguments_error
+  def run(options, _args, _config, %{token: nil}=endpoint) do
+    with_authentication(endpoint, &run(options, nil, nil, &1))
   end
```

Comment:
To clarify:

1. The action's `run` function gets called with an endpoint that has a `nil` token. 
2. The function head above matches against that and calls run again inside of a `with_authentication` wrapper.
3. The `with_authentication` wrapper fetches a token based on the user and password in the `endpoint`, merges the token into the `endpoint`, and calls `run` again with the merged endpoint.
4. Since the call from 3 has an endpoint with a token that isn't nil, we match against the second run function.

This lets us do things with the API in `run` before we dispatch to the `do_*` functions. 

I'm thinking it might be useful to lift this pattern out of the action altogether and elevate it to `Cogctl.Action` so we can do this across the board and remove the `with_authentication` bits from the actions entirely. That was out of scope though.